src/include/storage/checkpoint_manager.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include "common/spin_latch.h"
+#include "common/strong_typedef.h"
+#include "storage/checkpoint_io.h"
+#include "storage/projected_columns.h"
+#include "storage/sql_table.h"
+#include "storage/write_ahead_log/log_io.h"
+#include "transaction/transaction_context.h"
+#include "transaction/transaction_defs.h"
+
+namespace terrier::storage {
+
+/**
+ * A CheckpointManager is responsible for serializing tuples of all tables in the database out and
+ * divide time into epochs by doing checkpoint.
+ */
+class CheckpointManager {
```

Comment:
I'm not sure I understand how you envision the API being used.  The class is called `CheckpointManager` but it appears that all of the management (scheduling of next checkpoint, starting and stopping checkpoint-related transactions, tracking of tables to persist, etc.) is expected to occur somewhere else.  Semantically, I would expect the checkpoint manager to have a single entry point that manages the lifecycle of taking a checkpoint and getter/setter functions for adjusting its parameters (tables being persisted, checkpoint frequency, etc.).  Specifically, it should be expecting access to threads and the worker pool to manage these tasks internally rather than expecting another component to do this for you.

Reply:
You are correct. The checkpoint manager should 'manage' the process of checkpointing. However, we decided to first design the basic APIs for checkpoint and recovery, simply because there is no working catalog in the system. When there is a working catalog, a single call to StartCheckpoint() should first persist the catalog to the file. Then, for each user table, it will fetch its schema from the catalog and persist the table automatically.